
<h4 id="Université-de-Bordeaux,--Master-Mention-Informatique,--2021-2022">Université de Bordeaux,  Master Mention Informatique,  2021-2022<a class="anchor-link" href="#Université-de-Bordeaux,--Master-Mention-Informatique,--2021-2022">¶</a></h4><h1 id="Analyse,-classification-et-indexation-des-données:-feuille-4">Analyse, classification et indexation des données: feuille 4<a class="anchor-link" href="#Analyse,-classification-et-indexation-des-données:-feuille-4">¶</a></h1><h1 id="Machine-Learning">Machine Learning<a class="anchor-link" href="#Machine-Learning">¶</a></h1><h4 id="k-Nearest-Neighbors">k-Nearest Neighbors<a class="anchor-link" href="#k-Nearest-Neighbors">¶</a></h4>



<h2 id="Exercice-1.--k-Nearest-Neighbors-(from-scratch)">Exercice 1.  k-Nearest Neighbors (from scratch)<a class="anchor-link" href="#Exercice-1.--k-Nearest-Neighbors-(from-scratch)">¶</a></h2><p>Dans ce premier exercice, nous allons coder un algorithme de classification : le $k$-nn.</p>



<p>On commence par importer quelques modules python importants :</p>
<ul>
<li>Numpy : pour des calculs (algèbre linéaire, etc)</li>
<li>Pandas : pour la lecture des fichiers csv, etc </li>
</ul>


In [ ]:

import numpy as np
import pandas as pa

import warnings
warnings.filterwarnings("ignore")



In [ ]:
data = pa.read_csv("proprietes.csv")


<h4 id="Etape-1-:-Distance-euclidienne">Etape 1 : Distance euclidienne<a class="anchor-link" href="#Etape-1-:-Distance-euclidienne">¶</a></h4>



<ol>
<li>Ecrire une fonction pour calculer la distance (euclidienne) entre deux vecteurs. Attention, le vecteur donné comme deuxième paramètre de votre fonction contient, en dernière position, la classe de l'élément qu'il représente.</li>
</ol>


In [ ]:

def euclidean(u, v):
    return np.linalg.norm(u-v[:-1])




<p>Tester votre fonction. Les instructions suivantes devront produire le résultat de la cellule</p>


In [ ]:

u = np.array([1, 2, 3])
v = np.array([1, 2, 3, 1])
w = np.array([2, 3, 4, 0])
print(euclidean(u, v))
print(euclidean(u, w))


0.0
1.7320508075688772



<ol>
<li>Ecrire une fonction pour calculer les distances d'un vecteur à tous les autres vecteurs d'un dataset (matrice numpy - voir exemple).</li>
</ol>


In [ ]:

def distances(u, dataset):
    return [euclidean(u, i) for i in dataset]



<p>Tester votre fonction. Les instructions suivantes devront produire le résultat de la cellule</p>


In [ ]:
u = np.array([1, 2, 3])
dataset = np.array([[1, 2, 3, 0],
                   [2, 3, 4, 1],
                   [3, 4, 5, 1],
                   [4, 5, 6, 0]])
dist = distances(u, dataset)
print(dist)

[0.0, 1.7320508075688772, 3.4641016151377544, 5.196152422706632]



<h4 id="Etape-2-:-Récupérer-la-liste-des-$k$-voisins-les-plus-proches">Etape 2 : Récupérer la liste des $k$ voisins les plus proches<a class="anchor-link" href="#Etape-2-:-Récupérer-la-liste-des-$k$-voisins-les-plus-proches">¶</a></h4>



<ol>
<li>Ecrire une fonction <code>voisins</code> permettant de récupérer la liste des $k$-voisins d'un sommet les plus proches dans un dataset.</li>
</ol>


In [ ]:
def voisins(u,dataset,k):
    dists = []
    for ex in dataset:
        dists.append((ex,euclidean(u,ex)))
    dists.sort(key = lambda elt:elt[1])
    _voisins = []
    for i in range(k):
        _voisins.append(dists[i][0])
    return _voisins


<p>Tester votre fonction. Les instructions suivantes devront produire le résultat de la cellule.</p>


In [ ]:

u = np.array([1, 2, 3])
dataset = np.array([[1, 2, 3, 1],
                   [2, 3, 4, 1],
                   [3, 4, 5, 0],
                   [4, 5, 6, 0]])
print(voisins(u, dataset, 2))



[array([1, 2, 3, 1]), array([2, 3, 4, 1])]



<h4 id="Etape-3-:-Faire-des-prédictions">Etape 3 : Faire des prédictions<a class="anchor-link" href="#Etape-3-:-Faire-des-prédictions">¶</a></h4>



<ol>
<li>Ecrire une fonction <code> classifier()</code> retournant la classe d'un élément $u$.</li>
</ol>


In [ ]:

def classifier(u, dataset, k):
    listeVoisins = voisins(u,dataset,k)
    binList = []
    c0 = 0
    c1 = 0
    for voisin in listeVoisins:
        if(voisin[-1] == 0 ):
            c0 += 1
        else:
            c1 += 1
    if c0 > c1:
        return 0
    else:
        return 1


<p>Tester votre fonction. Les instructions suivantes devront produire le résultat de la cellule.</p>


In [ ]:

u = np.array([4, 5, 6])
dataset = np.array([[1, 2, 3, 0],
                   [2, 3, 4, 1],
                   [3, 4, 5, 0],
                   [4, 5, 6, 1],
                   [1, 2, 3, 0]])
#print(distances(u, dataset))
print(classifier(u, dataset, 2))



1



<h3 id="Exercice-2.-Application">Exercice 2. Application<a class="anchor-link" href="#Exercice-2.-Application">¶</a></h3>



<p>Dans cet exercice, nous allons appliquer l'algorithme écrit ci-dessus pour classifier des iris.
Pour cela, nous commençons par charger le datset :</p>


In [ ]:

from sklearn.datasets import load_iris
iris = load_iris()
print(iris.DESCR)



.. _iris_dataset:

Iris plants dataset
--------------------

**Data Set Characteristics:**

    :Number of Instances: 150 (50 in each of three classes)
    :Number of Attributes: 4 numeric, predictive attributes and the class
    :Attribute Information:
        - sepal length in cm
        - sepal width in cm
        - petal length in cm
        - petal width in cm
        - class:
                - Iris-Setosa
                - Iris-Versicolour
                - Iris-Virginica
                
    :Summary Statistics:

    ============== ==== ==== ======= ===== ====================
                    Min  Max   Mean    SD   Class Correlation
    ============== ==== ==== ======= ===== ====================
    sepal length:   4.3  7.9   5.84   0.83    0.7826
    sepal width:    2.0  4.4   3.05   0.43   -0.4194
    petal length:   1.0  6.9   3.76   1.76    0.9490  (high!)
    petal width:    0.1  2.5   1.20   0.76    0.9565  (high!)
    ============== ==== ==== ======= ===== ===========


<p>Comme indiqué dans sa description, le dataset contient, pour chaque iris, la longueur et la largeur de sa sépale et la longeur et la largeur de sa pétale. Les iris sont ensuite classifiés soit en Iris-Setosa (0), soit en Iris-Versicolour (1) ou encore en Iris-Virginica (2) :</p>


In [ ]:

print(iris.target)



[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2]



<p><b>Indication :</b> le dataset <code>iris</code> est composé de deux parties :</p>
<ul>
<li><code>iris.data</code> décrit les caractéristiques (features)</li>
<li><code>iris.target</code> contient les classes</li>
</ul>


In [ ]:

print(iris.data.shape)
print(iris.data)


(150, 4)
[[5.1 3.5 1.4 0.2]
 [4.9 3.  1.4 0.2]
 [4.7 3.2 1.3 0.2]
 [4.6 3.1 1.5 0.2]
 [5.  3.6 1.4 0.2]
 [5.4 3.9 1.7 0.4]
 [4.6 3.4 1.4 0.3]
 [5.  3.4 1.5 0.2]
 [4.4 2.9 1.4 0.2]
 [4.9 3.1 1.5 0.1]
 [5.4 3.7 1.5 0.2]
 [4.8 3.4 1.6 0.2]
 [4.8 3.  1.4 0.1]
 [4.3 3.  1.1 0.1]
 [5.8 4.  1.2 0.2]
 [5.7 4.4 1.5 0.4]
 [5.4 3.9 1.3 0.4]
 [5.1 3.5 1.4 0.3]
 [5.7 3.8 1.7 0.3]
 [5.1 3.8 1.5 0.3]
 [5.4 3.4 1.7 0.2]
 [5.1 3.7 1.5 0.4]
 [4.6 3.6 1.  0.2]
 [5.1 3.3 1.7 0.5]
 [4.8 3.4 1.9 0.2]
 [5.  3.  1.6 0.2]
 [5.  3.4 1.6 0.4]
 [5.2 3.5 1.5 0.2]
 [5.2 3.4 1.4 0.2]
 [4.7 3.2 1.6 0.2]
 [4.8 3.1 1.6 0.2]
 [5.4 3.4 1.5 0.4]
 [5.2 4.1 1.5 0.1]
 [5.5 4.2 1.4 0.2]
 [4.9 3.1 1.5 0.2]
 [5.  3.2 1.2 0.2]
 [5.5 3.5 1.3 0.2]
 [4.9 3.6 1.4 0.1]
 [4.4 3.  1.3 0.2]
 [5.1 3.4 1.5 0.2]
 [5.  3.5 1.3 0.3]
 [4.5 2.3 1.3 0.3]
 [4.4 3.2 1.3 0.2]
 [5.  3.5 1.6 0.6]
 [5.1 3.8 1.9 0.4]
 [4.8 3.  1.4 0.3]
 [5.1 3.8 1.6 0.2]
 [4.6 3.2 1.4 0.2]
 [5.3 3.7 1.5 0.2]
 [5.  3.3 1.4 0.2]
 [7.  3.2 4.7 1.4]
 [6.4 3.2 4.5 1.5]
 [6

In [ ]:

print(iris.target.shape)



(150,)



<p>Pour des raisons pédagogiques, et pour se focaliser sur l'algorithme, nous avons choisi de l'implémenter pour faire de la classification binaire (c'est ce que réalisent les étapes de l'exercice 1).</p>



<ol>
<li>Ecrire l'instruction permettant de transformer le problème en un problème de classification binaire : tous les Iris-Virginica (2) seront classés en Iris-Versicolour (1).</li>
</ol>


In [ ]:
iris.target = np.where(iris.target == 2 , 1,iris.target)
print(iris.target)


[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1]



<ol>
<li>Afin d'utiliser l'algorithme que vous avez implémenté dans l'exercice 1, créer une liste <code>dataset</code> dont le contenu sera des iris avec leur classes. Pour pouvoir visualiser le dataset, nous n'allons garder que les deux premières colonnes.</li>
</ol>


In [ ]:
#idataset = np.array([iris.data[:,:2],iris.target[:2]])
data = np.array(iris.data[:,:2])
target = np.array(iris.target[:])
dataset = np.array([np.append(data[i],target[i])for i in range(data.shape[0])])
print(dataset)

[[5.1 3.5 0. ]
 [4.9 3.  0. ]
 [4.7 3.2 0. ]
 [4.6 3.1 0. ]
 [5.  3.6 0. ]
 [5.4 3.9 0. ]
 [4.6 3.4 0. ]
 [5.  3.4 0. ]
 [4.4 2.9 0. ]
 [4.9 3.1 0. ]
 [5.4 3.7 0. ]
 [4.8 3.4 0. ]
 [4.8 3.  0. ]
 [4.3 3.  0. ]
 [5.8 4.  0. ]
 [5.7 4.4 0. ]
 [5.4 3.9 0. ]
 [5.1 3.5 0. ]
 [5.7 3.8 0. ]
 [5.1 3.8 0. ]
 [5.4 3.4 0. ]
 [5.1 3.7 0. ]
 [4.6 3.6 0. ]
 [5.1 3.3 0. ]
 [4.8 3.4 0. ]
 [5.  3.  0. ]
 [5.  3.4 0. ]
 [5.2 3.5 0. ]
 [5.2 3.4 0. ]
 [4.7 3.2 0. ]
 [4.8 3.1 0. ]
 [5.4 3.4 0. ]
 [5.2 4.1 0. ]
 [5.5 4.2 0. ]
 [4.9 3.1 0. ]
 [5.  3.2 0. ]
 [5.5 3.5 0. ]
 [4.9 3.6 0. ]
 [4.4 3.  0. ]
 [5.1 3.4 0. ]
 [5.  3.5 0. ]
 [4.5 2.3 0. ]
 [4.4 3.2 0. ]
 [5.  3.5 0. ]
 [5.1 3.8 0. ]
 [4.8 3.  0. ]
 [5.1 3.8 0. ]
 [4.6 3.2 0. ]
 [5.3 3.7 0. ]
 [5.  3.3 0. ]
 [7.  3.2 1. ]
 [6.4 3.2 1. ]
 [6.9 3.1 1. ]
 [5.5 2.3 1. ]
 [6.5 2.8 1. ]
 [5.7 2.8 1. ]
 [6.3 3.3 1. ]
 [4.9 2.4 1. ]
 [6.6 2.9 1. ]
 [5.2 2.7 1. ]
 [5.  2.  1. ]
 [5.9 3.  1. ]
 [6.  2.2 1. ]
 [6.1 2.9 1. ]
 [5.6 2.9 1. ]
 [6.7 3.1 1. ]
 [5.6 3.  


<ol>
<li>Visualiser votre dataset</li>
</ol>


In [ ]:
print(iris.feature_names)

['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']


In [ ]:
from matplotlib import pyplot as plt
from matplotlib import pyplot as plt
import seaborn as sns
sns.set()

X = dataset[:,0:1]
Y = dataset[:,1:2]
X = pa.DataFrame(iris.data[:,:2], columns=iris.feature_names[:2])
y = pa.Categorical.from_codes(iris.target, iris.target_names[:2])
y = pa.get_dummies(y, drop_first=True)

sns.scatterplot(
    x='sepal length (cm)',
    y='sepal width (cm)',
    hue='benign',
    data=X.join(y, how='outer')
)



ValueError: Could not interpret input 'benign'


<ol>
<li>Soit le vecteur $u(6.5, 2.5)$, utilisez votre algorithme avec $k=3$ pour classer $u$.</li>
</ol>


In [ ]:

classifier(np.array([6.5,2.5]),dataset,3)

1


<ol>
<li>Vérifier visuellement votre résultat.</li>
</ol>



<h3 id="Exercice-3-:-$k$-nn-avec-sklearn">Exercice 3 : $k$-nn avec <code>sklearn</code><a class="anchor-link" href="#Exercice-3-:-$k$-nn-avec-sklearn">¶</a></h3>



<p>Dans cet exercice, nous allons utiliser le classifieur $k$−nn pour apprendre à reconnaître des fruits. Pour cela, nous allons utiliser le dataset fruits disponible au format csv à l’adresse :</p>
<pre><code>           https://www.labri.fr/~zemmari/datasets/</code></pre>



<ol>
<li>Charger les données et afficher les informations pour vérifier si le dataset ne contient pas de données manquantes.</li>
</ol>



<ol>
<li>Afficher un graphique pour visualiser les tailles (height) en fonction des largeurs (width).</li>
</ol>



<ol>
<li>Modifier votre graphique pour qu’il affiche les points avec des couleurs différentes en fonction du
nom (<code>fruit_name</code>) du fruit. Vérifier que les classes sont assez équilibrées.</li>
</ol>



<ol>
<li>Définir X les données composées de la taille, la longueur et la masse des fruits, et Y les noms des
fruits.</li>
</ol>



<ol>
<li>Découper les données en deux parties : une pour l’entraînement et une pour le test.</li>
</ol>



<ol>
<li>Entraîner un classifieur $k$−nn à reconnaître les fruits. Mesurer les performances de votre classifieur. Quelle est la valeur par défaut de $k$ ?</li>
</ol>



<ol>
<li>Quelle est la nature du fruit dont la masse, la largeur et la taille sont données respectivement par 100, 6.3 et 8?</li>
</ol>



<ol>
<li>Trouver une "bonne" valeur pour k.</li>
</ol>
